File: adventofcode/year2017/Day11.cs
using adventofcode.common;
using adventofcode.common.grid;

namespace adventofcode.year2017;

public class Day11 : Solution
{
    private string _input;
    
    public Day11(string year, string day) : base(year, day)
    {
        _input = LoadInputAsString();
    }

    public override string PartOne()
    {
        var hg = new HexaGrid(_input);
        hg.Trace();
        
        return Convert.ToString(hg.Distance(hg.Position()));
    }

    public override string PartTwo()
    {
        var hg = new HexaGrid(_input);
        hg.Trace();
        
        return Convert.ToString(hg.MaxDistance());
    }

    private class HexaGrid
    {
        private List<string> _path;
        private Point2D _position;
        private int _max_distance;

        public HexaGrid(string path)
        {
            _path = new List<string>(path.Split(",").ToList());
            _position = new Point2D();
            _max_distance = 0;
        }

        public Point2D Position()
        {
            return _position;
        }

        public int MaxDistance()
        {
            return _max_distance;
        }

        public int Distance(Point2D position)
        {
            return Math.Abs(position.X() / 2) + Math.Abs(position.Y() / 2) + (Math.Abs(position.X() % 2) + Math.Abs(position.Y() % 2)) / 2;
        }

        private Point2D Offset(string direction) =>
            direction switch
            {
                "n" => new Point2D(0, -2),
                "s" => new Point2D(0, 2),
                "nw" => new Point2D(-1, -1),
                "ne" => new Point2D(1, -1),
                "sw" => new Point2D(-1, 1),
                "se" => new Point2D(1, 1),
            };

        public void Trace()
        {
            foreach (var step in _path)
            {
                _position += Offset(step);
                
                var distance = Distance(_position);
                _max_distance = distance > _max_distance ? distance : _max_distance;
            }
        }
    }
}

File: adventofcode/year2019/day11.py
from __future__ import annotations
from adventofcode.common import Solution
from adventofcode.common.grid import Point2D
from adventofcode.year2019.day09 import IntCodeCPUComplete
from typing import List

import sys


class HullPaintingRobot(object):
    def __init__(self, program: List[int], initial_input: int):
        self._cpu = IntCodeCPUComplete(program)
        self._cpu.add_input(initial_input)
        self._position = Point2D(0, 0)
        self._panels = {}
        self._orientation_index = 0
        self._deltas = (Point2D(0, -1), Point2D(1, 0), Point2D(0, 1), Point2D(-1, 0))   # up, right, down, left
        self._minx = sys.maxsize
        self._maxx = -sys.maxsize
        self._miny = sys.maxsize
        self._maxy = -sys.maxsize

    def run(self) -> int:
        painted = 0
        while not self._cpu.halted:
            # paint panel
            self._cpu.run()
            paint = self._cpu.get_output()
            painted += (1 if self._position not in self._panels.keys() else 0)
            self._panels[self._position] = paint

            # turn and move
            self._cpu.run()
            turn = self._cpu.get_output()
            match turn:
                case 0: # turn left
                    self._orientation_index = (self._orientation_index - 1) if (self._orientation_index - 1) >= 0 else (len(self._deltas) - 1)
                case 1: # turn right
                    self._orientation_index = (self._orientation_index + 1) if (self._orientation_index + 1) < len(self._deltas) else 0
                case _:
                    raise Exception(f"Invalid turn : {turn}")
            self._position += self._deltas[self._orientation_index]

            # update width and height tracking
            self._minx = self._position.x if self._position.x < self._minx else self._minx
            self._maxx = self._position.x if self._position.x > self._maxx else self._maxx
            self._miny = self._position.y if self._position.y < self._miny else self._miny
            self._maxy = self._position.y if self._position.y > self._maxy else self._maxy

            # feed next input
            self._cpu.add_input(self._panels[self._position] if self._position in self._panels.keys() else 0)

        return painted

    def show(self) -> None:
        for y in range(self._miny, self._maxy - self._miny + 1):
            row = []
            for x in range(self._minx, self._maxx - self._minx + 1):
                p = Point2D(x, y)
                paint = self._panels[p] if p in self._panels.keys() else 0
                row.append(' ' if paint == 0 else '#')
            print("".join(row))

class Day11(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)
        self._input = [int(i) for i in self._load_input_as_string().split(',')]

    def part_one(self):
        hpb = HullPaintingRobot(self._input, 0)
        return hpb.run()

    def part_two(self):
        hpb = HullPaintingRobot(self._input, 1)
        hpb.run()
        hpb.show()

        return ""
